coffeecard/CoffeeCard.Library/Services/v2/TokenService.cs

```diff
@@ -0,0 +1,66 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using CoffeeCard.Common.Errors;
+using CoffeeCard.Library.Persistence;
+using CoffeeCard.Models.Entities;
+using Microsoft.EntityFrameworkCore;
+
+namespace CoffeeCard.Library.Services.v2;
+
+public class TokenService : ITokenService
+{
+    private readonly CoffeeCardContext _context;
+    private readonly IHashService _hashService;
+
+    public TokenService(CoffeeCardContext context, IHashService hashService)
+    {
+        _context = context;
+        _hashService = hashService;
+    }
+
+    public async Task<string> GenerateMagicLink(User user)
+    {
+        var guid = Guid.NewGuid().ToString();
+        var magicLinkToken = new Token(guid, TokenType.MagicLink);
+
+        user.Tokens.Add(magicLinkToken);
+        await _context.SaveChangesAsync();
+        return magicLinkToken.TokenHash;
+    }
+
+    public async Task<string> GenerateRefreshTokenAsync(User user)
+    {
+        var refreshToken = Guid.NewGuid().ToString();
+        var hashedToken = _hashService.Hash(refreshToken);
+        _context.Tokens.Add(new Token(hashedToken, TokenType.Refresh) { UserId = user.Id });
+        await _context.SaveChangesAsync();
+        return refreshToken;
+    }
+
+    public async Task<Token> GetValidTokenByHashAsync(string tokenString)
+    {
+        var tokenHash = _hashService.Hash(tokenString);
+        var foundToken = await _context.Tokens.Include(t => t.User).FirstOrDefaultAsync(t => t.TokenHash == tokenHash);
+        if (foundToken == null || foundToken.Revoked || foundToken.Expired())
+        {
+            await InvalidateRefreshTokensForUser(foundToken?.User);
+            throw new ApiException("Invalid token", 401);
```

Comment:
Don't we have a more appropriate exception defined? I haven't checked, but I'd expect we had one

Reply:
Couldn't find one more appropriate. This is also how we handle failed authentication attempts in v1 `AccountService`.